common/flatpak-dir.c

```diff
@@ -8909,6 +8909,188 @@ flatpak_dir_deploy_update (FlatpakDir        *self,
+      if (g_str_has_prefix (icon_basename, old_app_id))
+        {
+          icon_suffix = icon_basename + strlen (old_app_id);
+          new_icon = g_strconcat (new_app_id, icon_suffix, NULL);
+          if (!g_file_set_display_name (icon_file, new_icon, NULL, &local_error))
+            {
+              g_warning ("Unable to rename icon file %s -> %s: %s", icon_basename, new_icon,
+                         local_error->message);
+            }
+        }
+    }
+}
+
+static void
+rewrite_dynamic_launchers (FlatpakDecomposed   *ref,
+                           const char * const * previous_ids)
+
+{
+  g_autoptr(GFile) portal_desktop_dir = NULL;
+  g_autofree char *portal_icon_path = NULL;
+  g_autofree char *app_id = NULL;
+  g_autoptr(GFileEnumerator) dir_enum = NULL;
+  g_autoptr(GError) local_error = NULL;
+
+  if (!flatpak_decomposed_is_app (ref))
+    return;
+
+  app_id = flatpak_decomposed_dup_id (ref);
+
+  /* Rename any dynamic launchers written by xdg-desktop-portal. The
+   * portal has its own code for renaming launchers on session start but we
+   * need to do it here as well so the launchers are correct in both cases:
+   * (1) the app rename transaction is being executed by the same user that
+   * has the launchers, or (2) the app is installed system-wide and another
+   * user has launchers.
+   */
+  if (previous_ids != NULL)
+    {
+      portal_desktop_dir = g_file_new_build_filename (g_get_user_data_dir (),
+                                                      "xdg-desktop-portal",
+                                                      "applications", NULL);
+      portal_icon_path = g_build_filename (g_get_user_data_dir (),
+                                           "xdg-desktop-portal", "icons", NULL);
+      dir_enum = g_file_enumerate_children (portal_desktop_dir,
+                                            G_FILE_ATTRIBUTE_STANDARD_NAME,
+                                            G_FILE_QUERY_INFO_NOFOLLOW_SYMLINKS,
+                                            NULL, &local_error);
+    }
+  if (dir_enum == NULL)
+    {
+      if (local_error && !g_error_matches (local_error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND))
+        {
+          g_warning ("Failed to enumerate portal desktop dir %s: %s",
+                     flatpak_file_get_path_cached (portal_desktop_dir),
+                     local_error->message);
+        }
+      g_clear_error (&local_error);
+    }
+  else
+    {
+      g_autoptr(GFileInfo) child_info = NULL;
+      while ((child_info = g_file_enumerator_next_file (dir_enum, NULL, &local_error)) != NULL)
+        {
+          const char *desktop_name;
+          int i;
+
+          desktop_name = g_file_info_get_name (child_info);
+          if (!g_str_has_suffix (desktop_name, ".desktop"))
+            continue;
+
+          for (i = 0; previous_ids[i] != NULL; i++)
+            {
+              if (g_str_has_prefix (desktop_name, previous_ids[i]))
```

Comment:
Are the `previous_ids` sorted by decreasing length? Otherwise there’s the possibility that the first match here will not be the longest prefix (barring other preconditions on `previous_ids` that I haven’t looked into). Is that OK?

Reply:
I think there is no guaranteed order for `previous_ids`, though in practice it would usually have only one entry, the app ID before the rename. This is not the same set of previous IDs that are written into the desktop file in `X-Flatpak-RenamedFrom` which is the whole history of previous IDs the app has had. I think the only case where this `previous_ids` array could have multiple entries would be if the app is renamed more than once and the local copy hasn't been updated for any of the renames (though I haven't checked if that possibility is definitely reachable with the codebase as it is). Then the only case where the sort order of `previous_ids` would matter would be if the app was renamed to something that was a suffix of one of its previous app IDs. So it's an edge case of an edge case but I guess you're right it's worth handling properly.

I can add something to sort the array by length.